File: src/infrastructure/validators/dates.ts
import { IValidationResult, ok, fail } from './result';

const DMY = /^\s*(\d{1,2})[/\-.](\d{1,2})[/\-.](\d{2,4})\s*$/;
const ISO = /^\s*(\d{4})-(\d{2})-(\d{2})\s*$/;

const MIN_YEAR = 1950;
const MAX_YEAR = 2100;

const expandYear = (y: number): number => {
  if (y >= 100) return y;
  return y <= 50 ? 2000 + y : 1900 + y;
};

const pad = (n: number): string => String(n).padStart(2, '0');

/** Converte uma data (dd/mm/aaaa ou ISO) para `YYYY-MM-DD`, validando existência e faixa. */
export function validateDate(raw: string | null | undefined): IValidationResult {
  const value = (raw ?? '').trim();
  if (!value) return fail('Data vazia');

  let year: number;
  let month: number;
  let day: number;

  const iso = ISO.exec(value);
  const dmy = DMY.exec(value);

  if (iso) {
    year = Number(iso[1]);
    month = Number(iso[2]);
    day = Number(iso[3]);
  } else if (dmy) {
    day = Number(dmy[1]);
    month = Number(dmy[2]);
    year = expandYear(Number(dmy[3]));
  } else {
    return fail(`Formato de data não reconhecido: "${value}"`);
  }

  if (year < MIN_YEAR || year > MAX_YEAR) return fail(`Ano fora da faixa permitida (${MIN_YEAR}-${MAX_YEAR}): ${year}`);
  if (month < 1 || month > 12) return fail(`Mês inválido: ${month}`);
  if (day < 1 || day > 31) return fail(`Dia inválido: ${day}`);

  // Valida data realmente existente (ex.: 31/02 não existe).
  const dt = new Date(Date.UTC(year, month - 1, day));
  if (dt.getUTCFullYear() !== year || dt.getUTCMonth() !== month - 1 || dt.getUTCDate() !== day) {
    return fail(`Data inexistente: "${value}"`);
  }

  return ok(`${year}-${pad(month)}-${pad(day)}`);
}

/** Garante que início <= fim (ambos já em ISO). */
export function validateRange(inicioIso: string | null, fimIso: string | null): IValidationResult {
  if (!inicioIso || !fimIso) return ok(''); // sem ambos, nada a comparar
  if (inicioIso > fimIso) return fail(`Data de início (${inicioIso}) posterior à data de término (${fimIso})`);
  return ok('');
}

File: app/validators/dates.py
"""Validação e normalização de datas (dd/mm/aaaa -> aaaa-mm-dd)."""
from __future__ import annotations

import datetime as _dt
import re

from .base import ValidationResult

# Faixa plausível de anos para datas de validade/registro.
MIN_YEAR = 1950
MAX_YEAR = 2100

_DMY = re.compile(r"^\s*(\d{1,2})[/\-.](\d{1,2})[/\-.](\d{2,4})\s*$")
_ISO = re.compile(r"^\s*(\d{4})-(\d{2})-(\d{2})\s*$")


def _normalize_year(year: int) -> int:
    """Expande anos de 2 dígitos para 4 (heurística de século)."""
    if year < 100:
        return 2000 + year if year <= 50 else 1900 + year
    return year


def validate(raw: str) -> ValidationResult:
    text = (raw or "").strip()
    if not text:
        return ValidationResult(False, None, ["Data vazia"])

    iso = _ISO.match(text)
    if iso:
        year, month, day = int(iso.group(1)), int(iso.group(2)), int(iso.group(3))
    else:
        match = _DMY.match(text)
        if not match:
            return ValidationResult(False, None, ["Formato de data não reconhecido"])
        day, month, year = int(match.group(1)), int(match.group(2)), _normalize_year(int(match.group(3)))

    if not (MIN_YEAR <= year <= MAX_YEAR):
        return ValidationResult(False, None, [f"Ano fora da faixa ({MIN_YEAR}-{MAX_YEAR}): {year}"])

    try:
        date = _dt.date(year, month, day)
    except ValueError:
        return ValidationResult(False, None, [f"Data inexistente: {day:02d}/{month:02d}/{year}"])

    return ValidationResult(True, date.isoformat(), [])


def validate_range(inicio: str, fim: str) -> ValidationResult:
    """Valida um intervalo de validade: início <= fim e ambos válidos."""
    v_inicio = validate(inicio)
    v_fim = validate(fim)
    messages: list[str] = []
    if not v_inicio.valid:
        messages.append("Início inválido: " + "; ".join(v_inicio.messages))
    if not v_fim.valid:
        messages.append("Fim inválido: " + "; ".join(v_fim.messages))
    if messages:
        return ValidationResult(False, None, messages)

    if v_inicio.normalized > v_fim.normalized:  # type: ignore[operator]
        return ValidationResult(
            False, None, ["Data de início posterior à data de fim"]
        )
    return ValidationResult(True, f"{v_inicio.normalized}..{v_fim.normalized}", [])
